nameresolution/sqlite/sqlite.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	s.db, err = sql.Open("sqlite", connString)
+	if err != nil {
+		return fmt.Errorf("failed to create connection: %w", err)
+	}
+
+	// Performs migrations
+	err = performMigrations(ctx, s.db, s.logger, migrationOptions{
+		HostsTableName:    s.metadata.TableName,
+		MetadataTableName: s.metadata.MetadataTableName,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to perform migrations: %w", err)
+	}
+
+	// Init the background GC
+	err = s.initGC()
+	if err != nil {
+		return err
+	}
+
+	// Register the host and update in background
+	err = s.registerHost(ctx)
+	if err != nil {
+		return err
+	}
+
+	go s.renewRegistration()
+
+	return nil
+}
+
+func (s *resolver) initGC() (err error) {
+	s.gc, err = internalsql.ScheduleGarbageCollector(internalsql.GCOptions{
+		Logger: s.logger,
+		UpdateLastCleanupQuery: func(arg any) (string, any) {
+			return fmt.Sprintf(`INSERT INTO %s (key, value)
+				VALUES ('nr-last-cleanup', CURRENT_TIMESTAMP)
+				ON CONFLICT (key)
+				DO UPDATE SET value = CURRENT_TIMESTAMP
+					WHERE (unixepoch(CURRENT_TIMESTAMP) - unixepoch(value)) * 1000 > ?;`,
+				s.metadata.MetadataTableName,
+			), arg
+		},
+		DeleteExpiredValuesQuery: fmt.Sprintf(
+			`DELETE FROM %s WHERE unixepoch(CURRENT_TIMESTAMP) - last_update < %d`,
+			s.metadata.TableName,
+			int(s.metadata.UpdateInterval.Seconds()),
+		),
+		CleanupInterval: s.metadata.CleanupInterval,
+		DB:              internalsql.AdaptDatabaseSQLConn(s.db),
+	})
+	return err
+}
+
+// Registers the host
+func (s *resolver) registerHost(ctx context.Context) error {
+	// Get the registration ID
+	u, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("failed to generate registration ID: %w", err)
+	}
+	s.registrationID = u.String()
+
+	queryCtx, queryCancel := context.WithTimeout(ctx, s.metadata.Timeout)
+	defer queryCancel()
+
+	// There's a unique index on address
+	// We use REPLACE to take over any previous registration for that address
+	_, err = s.db.ExecContext(queryCtx,
+		fmt.Sprintf("REPLACE INTO %s (registration_id, address, app_id, last_update) VALUES (?, ?, ?, unixepoch(CURRENT_TIMESTAMP))", s.metadata.TableName),
+		s.registrationID, s.metadata.GetAddress(), s.metadata.appID,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to register host: %w", err)
+	}
+
+	return nil
+}
+
+// In backgrounds, periodically renews the host's registration
+// Should be invoked in a background goroutine
+func (s *resolver) renewRegistration() {
+	addr := s.metadata.GetAddress()
+
+	d := s.metadata.UpdateInterval - s.metadata.Timeout
+	s.logger.Debugf("Started renewing host registration in background with interval %v", s.metadata.UpdateInterval)
+	t := time.NewTicker(d)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-s.closeCh:
+			// Component is closing
+			s.logger.Debug("Stopped renewing host registration: component is closing")
+			return
+
+		case <-t.C:
+			// Renew on the ticker
+			err := s.doRenewRegistration(context.Background(), addr)
```

Comment:
We should tie this context to `closeCh` else we could hang the system on close when request timeout is high.

Reply:
Will do